allow result of at::for_blob to advertise as resizeable (for tensor subclasses)

bdhirsh opened the pull request:
Previously, the first overload of `_make_wrapper_subclass` returned a tensor that **always** advertised as having a non-resizeable storage. Eventually, we'll need it be advertise as resizeable for functionalization to work (since functionalization occasionally needs to resize storages).

Not directly tested in this PR (tested more heavily later in aot dispatch, but if someone wants me to write a more direct test I can add one). 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #106405
* #106404
* #104482
* #107415
* __->__ #107416
* #107417

bdhirsh committed:
allow result of at::for_blob to advertise as resizeable (for tensor subclasses)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow result of at::for_blob to advertise as resizeable (for tensor subclasses)"

Previously, the first overload of `_make_wrapper_subclass` returned a tensor that **always** advertised as having a non-resizeable storage. Eventually, we'll need it be advertise as resizeable for functionalization to work (since functionalization occasionally needs to resize storages).

Not directly tested in this PR (tested more heavily later in aot dispatch, but if someone wants me to write a more direct test I can add one). 




[ghstack-poisoned]

albanD commented on aten/src/ATen/ops/from_blob.h:
```diff
@@ -85,6 +91,11 @@ class TORCH_API TensorMaker {
+  // Allows the storage of this tensor to be resized
+  // Note: this is won't actually resize anything (we use a meta allocator).
+  // This is only useful because at::for_blob is used by wrapper tensor subclasses,
+  // which we need to be able to advertise as resizeable.
```
I'm not sure about this comment? It seems to be written with the specific wrapper subclass use case of this API in mind.
"Allows the storage of the created tensor to be resized" sounds like an accurate description to me.
Since resizeable is a known concept for Tensor/Storage, I'm not sure that warrants a comment here.

ezyang replied:
So, the problem with resizing an `from_blob` tensor is we don't know how the original tensor was allocated, and so we cannot reallocate it in an equivalent way. This can matter if, e.g., the tensor is on shared/pinned memory.  For meta storage, however, it doesn't really matter, because there is no nontrivial allocator involved.

I guess you can kind of use this API if you don't mind us using the default allocator to service your resize. This might be the most reasonable semantics for this outside of Meta too.

bdhirsh replied:
I originally planned on making this more limited - `resizeable_storage()` will **never** let you use a real allocator, and is just useful for flipping metadata (important for subclasses).

Is the argument that we want to something like `resizeable_storage(/*meta=*/true/false)`, so real users of `for_blob()` can indicate that they want us to attempt to use our default allocator? I hesitated to do that since there isn't a concrete use case that I know of, while the meta-allocator version has a more concrete use case. But open to either

albanD replied:
In that case, should we request an allocator to be able to set resizable? And it is the user responsability to privide both.
We can then make _make_wrapper_subclass pass the Meta allocator as we know it is what we want there.

bdhirsh replied:
sure - so concretely, instead of a `.resizeable_storage()` API, we have a `.allocator(my_alloc)` API that lets you pass in a custom allocator, where that allocator's resizeable-ness is what determines whether or not our storage will be resizeable? And I'll pass the meta allocator in from `_make_wrapper_subclass`.

albanD replied:
You will need 2 APIs actually: one to pass in an allocator and one to say if this particular Tensor is resizable or not.

ezyang replied:
There's not much point in passing an allocator if it's not resizable though 📦

albanD commented on aten/src/ATen/templates/Functions.cpp:
```diff
@@ -36,7 +37,7 @@ Tensor TensorMaker::make_tensor() {
    }
 
-   Storage storage{Storage::use_byte_size_t{}, size_bytes, std::move(data_ptr)};
+   Storage storage{Storage::use_byte_size_t{}, size_bytes, std::move(data_ptr), /*allocator=*/c10::GetAllocator(c10::kMeta), /*resizeable=*/resizeable_};
```
A nullptr allocator is not the same as Meta allocator... 
I'm not sure what is the impact of always giving an allocator here tbh.
Should we only pass one if resizable_ is true to not change current behavior?

bdhirsh replied:
> Should we only pass one if resizable_ is true to not change current behavior?

yep sounds good!

bdhirsh committed:
Update on "allow result of at::for_blob to advertise as resizeable (for tensor subclasses)"

Previously, the first overload of `_make_wrapper_subclass` returned a tensor that **always** advertised as having a non-resizeable storage. Eventually, we'll need it be advertise as resizeable for functionalization to work (since functionalization occasionally needs to resize storages).

Not directly tested in this PR (tested more heavily later in aot dispatch, but if someone wants me to write a more direct test I can add one). 




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

bdhirsh committed:
Update on "allow result of at::for_blob to advertise as resizeable (for tensor subclasses)"

Previously, the first overload of `_make_wrapper_subclass` returned a tensor that **always** advertised as having a non-resizeable storage. Eventually, we'll need it be advertise as resizeable for functionalization to work (since functionalization occasionally needs to resize storages).

Not directly tested in this PR (tested more heavily later in aot dispatch, but if someone wants me to write a more direct test I can add one). 




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge
